### PR TITLE
internal/xds: convert v3 protobufs automatically

### PR DESCRIPTION
**Pull request by jpeach**
Signed-off-by: James Peach <jpeach@vmware.com>

### CODE REVIEW

**Comment by stevesloka** [comment]
Thanks for this @jpeach! This is a neat approach to moving form one version to the next. Taking some time to think through this a bit more and how it might integrate I'm leaning towards our original plan of duplicating code into a `v3` package. 

- One goal was to allow users to decide if they want to run a v2 or v3 cluster, so doing this approach won't let us serve v2 once we start migrating over to v3 unless, like you mentioned, we'd need to convert from v3 back to v2. 
- There is a deprecated field with that I'm aware of already (`DrainConnectionsOnHostRemoval` has been removed from v3), so we'd have a special case there to convert. I'm not sure if there are more or not. 

I'm still open to discussing this moving forward since this is a pretty big decision. What is everyone else's thoughts?

File: internal/xds/util.go [hunk 1]
```diff
@@ -38,3 +60,157 @@ func ClusterLoadAssignmentName(service types.NamespacedName, portName string) st
...
+		}
+
+		for _, f := range v3.HttpFilters {
+			if c := f.GetTypedConfig(); c != nil {
+				f.ConfigType = &envoy_extensions_filters_network_http_connection_manager_v3.HttpFilter_TypedConfig{
+					TypedConfig: Fixup(c).(*any.Any),
+				}
+			}
+		}
+
```
Comment (skriss) [inline internal/xds/util.go:163]: I think you need something like this in here for the access log config, it was still coming through as v2 without:

```
for i, a := range v3.AccessLog {
        v3.AccessLog[i].ConfigType = &envoy_config_accesslog_v3.AccessLog_TypedConfig{
	        TypedConfig: Fixup(a.GetTypedConfig()).(*any.Any),
	}
}
```

**Comment by jpeach** [comment]
I verified that migrating the protobufs through JSON serialization drops the migrated fields:

```
2020/10/30 10:47:01 original *envoy_api_v2.Cluster:
{"name":"default/httpbin/80/f7b4732cd5","altStatName":"default_httpbin_80","type":"EDS","edsClusterConfig":{"edsConfig":{"apiConfigSource":{"apiType":"GRPC","grpcServices":[{"envoyGrpc":{"clusterName":"contour"}}]}},"serviceName":"default/httpbin/http"},"connectTimeout":"0.250s","healthChecks":[{"timeout":"2s","interval":"10s","unhealthyThreshold":3,"healthyThreshold":2,"httpHealthCheck":{"host":"contour-envoy-healthcheck","path":"/status/200"}}],
"commonLbConfig":{"healthyPanicThreshold":{}},"drainConnectionsOnHostRemoval":true}
```
```
2020/10/30 10:47:01 converted *envoy_config_cluster_v3.Cluster:
{"name":"default/httpbin/80/f7b4732cd5","altStatName":"default_httpbin_80","type":"EDS","edsClusterConfig":{"edsConfig":{"apiConfigSource":{"apiType":"GRPC","transportApiVersion":"V3","grpcServices":[{"envoyGrpc":{"clusterName":"contour"}}]},"resourceApiVersion":"V3"},"serviceName":"default/httpbin/http"},"connectTimeout":"0.250s","healthChecks":[{"timeout":"2s","interval":"10s","unhealthyThreshold":3,"healthyThreshold":2,"httpHealthCheck":{"host":"contour-envoy-healthcheck","path":"/status/200"}}],
"commonLbConfig":{"healthyPanicThreshold":{}}}
```

Envoy has internal code that migrates field in the version converter. Possibly if we just rewrite the typeURL the internal envoy code will do the right thing. Experimenting ...

**Comment by jpeach** [comment]
> Envoy has internal code that migrates field in the version converter. Possibly if we just rewrite the typeURL the internal envoy code will do the right thing. Experimenting ...

Confirmed that this works:

```
"cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "default/httpbin/80/f7b4732cd5",
      "type": "EDS",
      "eds_cluster_config": {
       "eds_config": {
        "api_config_source": {
         "api_type": "GRPC",
         "grpc_services": [
          {
           "envoy_grpc": {
            "cluster_name": "contour"
           }
          }
         ]
        }
       },
       "service_name": "default/httpbin/http"
      },
      "connect_timeout": "0.250s",
      "health_checks": [
       {
        "timeout": "2s",
        "interval": "10s",
        "unhealthy_threshold": 3,
        "healthy_threshold": 2,
        "http_health_check": {
         "host": "contour-envoy-healthcheck",
         "path": "/status/200"
        }
       }
      ],
      "common_lb_config": {
       "healthy_panic_threshold": {}
      },
      "alt_stat_name": "default_httpbin_80",
      "ignore_health_on_host_removal": true
     },
```

File: internal/xds/util.go (hunk 1 above)
Comment (jpeach) [reply internal/xds/util.go:197]: fixed in HCM typeswitch

File: internal/xds/util.go [hunk 2]
```diff
@@ -38,3 +61,158 @@ func ClusterLoadAssignmentName(service types.NamespacedName, portName string) st
 
 	return strings.Join(name, "/")
 }
+
+// TypeMapping maps xDS type URLs from v2 to v3.
+var TypeMapping map[string]string
+
+func init() {
```
Comment (youngnick) [inline internal/xds/util.go:68]: Nit: Why do this one on `init()` rather than statically? Maybe might be worth a comment as well for future us.

**Review by youngnick — APPROVED** [review]
LGTM with one small comment; I think @stevesloka needs to approve too though, as I'm not familiar enough with the protos to be 100%.

File: internal/xds/util.go (hunk 2 above)
Comment (jpeach) [reply internal/xds/util.go:70]: Originally I had something like `var TypeMapping = build_table()`. I switched to `init`, but I don't think there's much practical difference. Happy to take whichever approach is preferred :)

File: internal/xds/util.go [hunk 3]
```diff
@@ -38,3 +63,161 @@ func ClusterLoadAssignmentName(service types.NamespacedName, portName string) st
...
+
+	case *envoy_api_auth_v2.UpstreamTlsContext:
+		for _, s := range msg.GetCommonTlsContext().TlsCertificateSdsSecretConfigs {
+			rewriteConfigSource(s.GetSdsConfig())
+		}
+
+		return msg
+
+	default:
+		log.Printf("missing conversion for %T, good luck", msg)
```
Comment (stevesloka) [inline internal/xds/util.go:220]: Add the context where this is logging and remove `good luck`, it's not needed.

**Review by stevesloka — CHANGES_REQUESTED** [review]
Testing through this on my local cluster I keep getting log messages like this:

```
2020/10/30 10:42:12 missing conversion for *envoy_config_accesslog_v2.FileAccessLog, good luck
2020/10/30 10:42:12 missing conversion for *envoy_config_filter_http_lua_v2.Lua, good luck
2020/10/30 10:42:12 missing conversion for *envoy_config_filter_http_ext_authz_v2.ExtAuthz, good luck
2020/10/30 10:42:12 missing conversion for *envoy_config_accesslog_v2.FileAccessLog, good luck
2020/10/30 10:42:12 missing conversion for *envoy_config_filter_http_ext_authz_v2.ExtAuthzPerRoute, good luck
2020/10/30 10:42:12 missing conversion for *envoy_extensions_filters_network_http_connection_manager_v3.HttpConnectionManager, good luck
```

File: internal/xds/util.go (hunk 3 above)
Comment (skriss) [reply internal/xds/util.go:220]: If we're gonna log anything here (not sure we need to once this is finalized), we should plumb a logrus logger through

File: internal/xds/v3/contour.go [hunk 4]
```diff
@@ -114,15 +119,17 @@ func (s *contourServer) stream(st grpcStream) error {
...
 		}
 
-		// from the request we derive the resource to stream which have
+		// From the request we derive the resource to stream which have
 		// been registered according to the typeURL.
-		r, ok := s.resources[req.TypeUrl]
+		r, ok := s.resources[req.GetTypeUrl()]
 		if !ok {
-			return done(log, fmt.Errorf("no resource registered for typeURL %q", req.TypeUrl))
+			if !ok {
```
Comment (skriss) [inline internal/xds/v3/contour.go:126]: remove extraneous `if !ok {`

File: internal/xds/util.go (hunk 2 above)
Comment (jpeach) [reply internal/xds/util.go:69]: I'll leave this as it for now; we will remove the code in Contour 1.11 when we are fully over to v3.

**Review by skriss — APPROVED** [review]
No other comments; LGTM assuming testing's OK (has seemed fine in my limited smoke-tests).

**Pull request merged**